### PR TITLE
interfaces: add desktop-file-ids plug attr to desktop interface

### PR DESCRIPTION
**Pull request by ZeyadYasser**
This plug attribute is needed to allow snaps to have custom desktop file names.

Previously, snapd installed desktop files as `$snap_$app.desktop`. This causes problems in a number of cases where desktop files are expected to follow a particular format, or have a specific name.

The new plug attribute allows snaps to have custom desktop file names listed under the `desktop-file-ids` attribute.

Naming cannot be left uncontrolled, so the `$MISSING` constraint is added to only allow overriding `desktop-file-ids` through the store. The process of requesting the snap declaration on the store will help ensure that the prefixes are sensible.

More details can be found in the spec: [SD170](https://docs.google.com/document/d/1j4flP303OJKrSxSCjkeZkh-9COuahVmo_EK_ye09RxQ/edit)

**Note: deny-auto-connection and deny-connection had to be moved to the plug base declaration because it takes precedence as mentioned in [interfaces/builtin/README.md](https://github.com/snapcore/snapd/blob/master/interfaces/builtin/README.md#base-declaration-policy-patterns)**

### CODE REVIEW

**Comment by ZeyadYasser** [comment]
@jhenstridge It would be great if you could take a look, I wasn't able to add you as a reviewer after the org migration, so pinging here instead.

File: interfaces/policy/basedeclaration_test.go [hunk 1]
```diff
@@ -1082,6 +1082,27 @@ func (s *baseDeclSuite) TestPlugInstallation(c *C) {
...
+	ic := s.installPlugCand(c, "desktop", snap.TypeApp, `name: desktop
+version: 0
+type: app
+plugs:
+  desktop:
+`)
+	err := ic.Check()
+	c.Assert(err, IsNil)
+	// desktop-file-ids can only be set by the store
+	ic = s.installPlugCand(c, "desktop", snap.TypeApp, `name: desktop
```
Comment (MiguelPires) [inline interfaces/policy/basedeclaration_test.go:1096]: nitpick: if this doesn't fit in the test's pattern, I'd prefer to put it in its own test to avoid having each test do too much and become hard to understand

**Review by MiguelPires — APPROVED** [review]
lgtm

File: interfaces/builtin/desktop.go [hunk 2]
```diff
@@ -609,9 +616,37 @@ func (iface *desktopInterface) AppArmorPermanentSlot(spec *apparmor.Specificatio
...
+	desktopFileIDs, ok := attrVal.([]interface{})
+	if !ok {
+		return fmt.Errorf(`cannot add %s plug: "desktop-file-ids" must be a list of strings`, iface.name)
+	}
+	for _, desktopFileID := range desktopFileIDs {
+		if _, ok := desktopFileID.(string); !ok {
+			return fmt.Errorf(`cannot add %s plug: "desktop-file-ids" must be a list of strings`, iface.name)
+		}
+	}
+
```
Comment (alexmurray) [inline interfaces/builtin/desktop.go:636]: Should `desktopFileIDs` be further validated as per https://specifications.freedesktop.org/desktop-entry-spec/latest/file-naming.html ? Whilst I understand that a snap declaration is required for the use of any entries here, I wonder if we should have more of a defence-in-depth approach and not just rely on the store for this validation but instead have snapd validate these?

**Review by alexmurray — CHANGES_REQUESTED** [review]
I think this should be extended to validate the specified desktop file ids as per the Freedesktop specification at https://specifications.freedesktop.org/desktop-entry-spec/latest/file-naming.html

File: interfaces/builtin/desktop.go (hunk 2 above)
Comment (ZeyadYasser) [reply interfaces/builtin/desktop.go:651]: Great point, Thanks for pointing this out! I updated the validation logic and added a corresponding test.

File: interfaces/policy/basedeclaration_test.go (hunk 1 above)
Comment (ZeyadYasser) [reply interfaces/policy/basedeclaration_test.go]: Thanks for pointing this out, It made me reconsider and created a separate test that actually tests the store-side snap-declaration override.

File: interfaces/builtin/desktop.go [hunk 3]
```diff
@@ -609,9 +617,51 @@ func (iface *desktopInterface) AppArmorPermanentSlot(spec *apparmor.Specificatio
...
 
+// https://specifications.freedesktop.org/desktop-entry-spec/latest/file-naming.html
+// Desktop file id must be a valid D-Bus name:
+//  - A sequence of non-empty elements separated by dots
+//  - None of which starts with a digit
+//  - Each of which contains only characters from the set [A-Za-z0-9-_]
+//
+// XXX: dashes "-" are not recommended but supported, should they be removed?
+// https://dbus.freedesktop.org/doc/dbus-specification.html#message-protocol-names
+var desktopFileIDRegexp = regexp.MustCompile("^([A-Za-z-_][\\w-]*)(\\.[A-Za-z-_][\\w-]*)*$")
```
Comment (valentindavid) [inline interfaces/builtin/desktop.go:628]: Use a raw string instead of escaping the backslashes.

Comment (valentindavid) [reply interfaces/builtin/desktop.go:628]: Also I see that the specification uses `[A-Za-z0-9-_]`. But in my opinion, it is is better to put the dash when it is not used for a range as the last character of a class. Like `[A-Za-z0-9_-]`. Because without parsing it from left to right, you can know for sure the dash here is is included in the class.

**Review by valentindavid — APPROVED** [review]
Looks good to me.

File: interfaces/builtin/desktop.go (hunk 3 above)
Comment (ZeyadYasser) [reply interfaces/builtin/desktop.go]: Thanks, updated!

**Review by alexmurray — APPROVED** [review]
LGTM!

File: interfaces/builtin/desktop.go [hunk 4]
```diff
@@ -609,9 +617,51 @@ func (iface *desktopInterface) AppArmorPermanentSlot(spec *apparmor.Specificatio
 	return nil
 }
```
Comment (pedronis) [inline interfaces/builtin/desktop.go:618]: it's probably a good idea to add specific tests about tthe "deny-*connection" rules here now that I have moved to the plug side

**Review by pedronis — COMMENTED** [review]
it's probably a good idea to add more tests now about the base declaration rules, basically we need to try to cover both the positive and negative cases, atm it seems we have only 

baseDeclSuite.TestAutoConnectionImplicitSlotOnly

File: interfaces/builtin/desktop.go (hunk 4 above)
Comment (ZeyadYasser) [reply interfaces/builtin/desktop.go:618]: There is a test for slot "allow/deny-installation" in baseDeclSuite.TestSlotInstallation
https://github.com/canonical/snapd/blob/af81a100792825a10430c461416896f4c3e70b3d/interfaces/policy/basedeclaration_test.go#L924-L935
There is also a test for plug-side "allow-installation" in baseDeclSuite.TestDesktopFileIDsOverride.

I will now add specific tests for deny* rule, thanks for pointing that out.

Comment (ZeyadYasser) [reply interfaces/builtin/desktop.go:618]: Done

**Review by pedronis — APPROVED** [review]
thank you

**Comment by 3v1n0** [comment]
Thanks for start working on this, one aspect I'm also wondering we should consider is that a snap should be only allowed to have such interface *only if* one of its apps exposes the such desktop file ID interface as the common ID (as per https://github.com/canonical/snapd/commit/360469e6c07ade130de3e762dbd068b1b48ac3ae)

Now, as I'm not sure how a such application will be exposed externally after this change (if both desktop files will be created or not), but `CommonID` is expected to match such ID anyways.

**Comment by ZeyadYasser** [comment]
> Thanks for start working on this, one aspect I'm also wondering we should consider is that a snap should be only allowed to have such interface _only if_ one of its apps exposes the such desktop file ID interface as the common ID (as per [360469e](https://github.com/canonical/snapd/commit/360469e6c07ade130de3e762dbd068b1b48ac3ae))
> 
> Now, as I'm not sure how a such application will be exposed externally after this change (if both desktop files will be created or not), but `CommonID` is expected to match such ID anyways.

I don't have strong opinions on enforcing this restriction but maybe @pedronis  @jhenstridge have some insights to share.

**Comment by 3v1n0** [comment]
> I don't have strong opinions on enforcing this restriction but maybe @pedronis @jhenstridge have some insights to share.

More than at snap level such restriction should be at global level though, I've not checked all the details, but for sure there should be not two snaps registering the same desktop-id, but two different snaps can expose the plug (like two different versions of the same app, or two different packaging of it) so that only one of them uses it.
